File: apps/windows/TeslaSync.App/modals-dialogs/CommandSelectDialog.Model.cs
using System.Globalization;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.ModalsDialogs;

/// <summary>
/// One selectable option in the <c>CommandSelectDialog</c> — the native analogue of the web <c>SelectOption</c>
/// (web/src/features/system/commands.ts). <see cref="LabelKey"/> / <see cref="LabelFallback"/> resolve through the
/// i18n facade (web <c>t(opt.labelKey, opt.labelFallback)</c>); <see cref="Description"/> is the optional, already
/// human-readable sub-line rendered verbatim (web renders <c>opt.description</c> directly, not through <c>t()</c>).
/// <see cref="Value"/> is the command-parameter payload emitted when the option is chosen (web <c>onSelect(opt.value)</c>).
/// </summary>
public sealed record CommandSelectOption(string Value, string LabelKey, string LabelFallback, string? Description = null);

/// <summary>
/// The command-select request the dialog renders — the native analogue of the web <c>def</c> +
/// <c>def.selectConfig</c> props of <c>CommandSelectDialog</c> (web/src/features/system/components/CommandSelectDialog.tsx).
/// It carries the keyed title (web <c>t(def.labelKey, def.labelFallback)</c>), the leading icon glyph (web
/// <c>def.icon</c>), the parameter name the chosen value populates (web <c>sc.paramName</c>) and the list of
/// <see cref="CommandSelectOption"/>s. The view never fetches this — the parent (web Vehicle Commands page) supplies it.
/// </summary>
public sealed class CommandSelectRequest
{
    /// <summary>Creates the request over its keyed title, parameter name, options and optional icon glyph.</summary>
    /// <param name="titleKey">The i18n key for the dialog title (web <c>def.labelKey</c>).</param>
    /// <param name="titleFallback">The English fallback for the title (web <c>def.labelFallback</c>).</param>
    /// <param name="paramName">The command parameter the chosen value populates (web <c>sc.paramName</c>).</param>
    /// <param name="options">The selectable options (web <c>sc.options</c>); null entries are dropped.</param>
    /// <param name="iconGlyph">Optional Segoe Fluent glyph for the leading icon (web <c>def.icon</c>).</param>
    public CommandSelectRequest(
        string titleKey,
        string titleFallback,
        string paramName,
        IEnumerable<CommandSelectOption>? options,
        string? iconGlyph = null)
    {
        TitleKey = titleKey ?? string.Empty;
        TitleFallback = titleFallback ?? string.Empty;
        ParamName = paramName ?? string.Empty;
        Options = CommandSelectProjection.Normalize(options);
        IconGlyph = string.IsNullOrWhiteSpace(iconGlyph) ? CommandSelectRegistration.DefaultIconGlyph : iconGlyph!;
    }

    /// <summary>The i18n key for the dialog title (web <c>def.labelKey</c>).</summary>
    public string TitleKey { get; }

    /// <summary>The English fallback for the title (web <c>def.labelFallback</c>).</summary>
    public string TitleFallback { get; }

    /// <summary>The command parameter the chosen value populates (web <c>sc.paramName</c>).</summary>
    public string ParamName { get; }

    /// <summary>The normalized (null-dropped, order-preserving) list of options (web <c>sc.options</c>).</summary>
    public IReadOnlyList<CommandSelectOption> Options { get; }

    /// <summary>The leading icon glyph (web <c>def.icon</c>); defaults to <see cref="CommandSelectRegistration.DefaultIconGlyph"/>.</summary>
    public string IconGlyph { get; }
}

/// <summary>
/// A display-ready option — the native projection of a <see cref="CommandSelectOption"/> with its label resolved
/// through the i18n facade. The view binds these directly so it never resolves keys or recomputes the
/// description-visibility gate inline (web <c>{opt.description &amp;&amp; …}</c>).
/// </summary>
public sealed record CommandSelectResolvedOption(string Value, string Label, string? Description)
{
    /// <summary>True when a description sub-line should render (web truthiness <c>opt.description &amp;&amp;</c>: empty hides).</summary>
    public bool HasDescription => !string.IsNullOrEmpty(Description);
}

/// <summary>
/// Canonical slug, default icon and i18n keys for the <c>CommandSelectDialog</c> surface — the native mirror of
/// <c>web/src/features/system/components/CommandSelectDialog.tsx</c>. The only literal copy intrinsic to the web
/// component is the Cancel label (<c>t('common.cancel', 'Cancel')</c>); the title and per-option labels are
/// caller-supplied dynamic keys resolved through the i18n facade. The empty-state copy has no web counterpart (the
/// web maps over a non-empty option list) and is keyed here so the native empty branch is never a blank box. UI-free
/// so every key, fallback and bound is asserted headlessly.
/// </summary>
public static class CommandSelectRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "CommandSelectDialog";

    /// <summary>Default leading icon glyph (Segoe Fluent — <c>Repair/Options</c>) when the request supplies none.</summary>
    public const string DefaultIconGlyph = "\uE71C";

    /// <summary>Cancel button label (web <c>t('common.cancel', 'Cancel')</c>).</summary>
    public static string CancelLabel(ILocalizer localizer) =>
        Require(localizer).GetString("common.cancel", "Cancel");

    /// <summary>
    /// Friendly empty-state message shown when the request carries no options. The web component always renders a
    /// non-empty option list, so this copy is native-only and keyed (with its English fallback) so the empty branch
    /// resolves through the i18n facade rather than a hardcoded literal.
    /// </summary>
    public static string EmptyMessage(ILocalizer localizer) =>
        Require(localizer).GetString("commands.select.noOptions", "No options available");

    /// <summary>Resolve the dialog title from a request (web <c>t(def.labelKey, def.labelFallback)</c>).</summary>
    public static string Title(ILocalizer localizer, CommandSelectRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);
        return Require(localizer).GetString(request.TitleKey, request.TitleFallback);
    }

    /// <summary>Resolve a single option's label (web <c>t(opt.labelKey, opt.labelFallback)</c>).</summary>
    public static string OptionLabel(ILocalizer localizer, CommandSelectOption option)
    {
        ArgumentNullException.ThrowIfNull(option);
        return Require(localizer).GetString(option.LabelKey, option.LabelFallback);
    }

    private static ILocalizer Require(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer;
    }
}

/// <summary>
/// Pure projections for the <c>CommandSelectDialog</c> surface — the native analogue of the web component's
/// <c>sc.options.map(…)</c> render: dropping unusable entries, resolving each label through the i18n facade and
/// detecting the empty branch. UI-free so the option-resolution and empty-gate are unit-tested headlessly and the
/// view-model never recomputes them inline.
/// </summary>
public static class CommandSelectProjection
{
    /// <summary>
    /// Normalize the supplied options to the renderable list — null input becomes empty and null entries are
    /// dropped (null safety), while order and every non-null option are preserved (web maps the list as given).
    /// </summary>
    public static IReadOnlyList<CommandSelectOption> Normalize(IEnumerable<CommandSelectOption>? options)
    {
        if (options is null)
        {
            return Array.Empty<CommandSelectOption>();
        }

        var list = new List<CommandSelectOption>();
        foreach (var option in options)
        {
            if (option is not null)
            {
                list.Add(option);
            }
        }

        return list;
    }

    /// <summary>True when there is nothing to choose from (the native empty branch); web never hits this.</summary>
    public static bool IsEmpty(IReadOnlyList<CommandSelectOption>? options) => (options?.Count ?? 0) == 0;

    /// <summary>Resolve one option for display (web <c>t(opt.labelKey, …)</c> + the verbatim description).</summary>
    public static CommandSelectResolvedOption Resolve(ILocalizer localizer, CommandSelectOption option)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        ArgumentNullException.ThrowIfNull(option);
        return new CommandSelectResolvedOption(
            option.Value ?? string.Empty,
            CommandSelectRegistration.OptionLabel(localizer, option),
            option.Description);
    }

    /// <summary>Resolve every option for display, preserving order (web <c>sc.options.map(…)</c>).</summary>
    public static IReadOnlyList<CommandSelectResolvedOption> ResolveAll(
        ILocalizer localizer,
        IEnumerable<CommandSelectOption>? options)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        var resolved = new List<CommandSelectResolvedOption>();
        foreach (var option in Normalize(options))
        {
            resolved.Add(Resolve(localizer, option));
        }

        return resolved;
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>CommandSelectDialog</c> surface (P1/S11 diagnostics contract). Records only
/// operational counters with the surface slug — never the chosen command value or option labels — so a diagnostics
/// line can never leak which command a user issued. Thread-safe.
/// </summary>
public sealed class CommandSelectDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _optionsSelected;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public CommandSelectDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of option selections emitted from this surface.</summary>
    public long OptionsSelected => Interlocked.Read(ref _optionsSelected);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=CommandSelectDialog</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"view.opened slug={CommandSelectRegistration.Slug}"));
    }

    /// <summary>Record that an option was chosen (the value / label are never logged).</summary>
    public void RecordOptionSelected()
    {
        Interlocked.Increment(ref _optionsSelected);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"command.optionSelected slug={CommandSelectRegistration.Slug}"));
    }
}

File: apps/android/app/src/main/java/com/teslasync/modals-dialogs/CommandSelectDialog/CommandSelectDialogModel.kt
// Pure, framework-free model + projection for the CommandSelectDialog modal/dialog surface — the native analogue of
// everything the web component derives before it returns JSX (web/src/features/system/components/CommandSelectDialog.tsx).
// No Compose, no Android, no HTTP: every declaration here is exercised off-device by the :android:testReleaseUnitTest
// gate, so the composable stays a thin render layer over these pure functions.
//
// The web component is the option picker raised from a "select"-type vehicle command tile (e.g. seat-heater level,
// steering-wheel-heat level): a *controlled* dialog whose only data dependency is `useTranslation` (i18n, P1/S10). It
// binds no fetch and owns no store — the option list arrives pre-built in the `def: CommandDef` prop (the page's static
// `commands.ts` config), the owner decides when it is `open`, and the chosen option's value is handed straight back to
// the owner through `onSelect(opt.value)`. So (exactly like the sibling ConfirmDialog / AcknowledgeAlertDialog surfaces)
// the cache-then-network lifecycle (loading / empty / error / stale / offline) belongs to the OWNING command surface
// that decides to raise this picker, NOT here; modelling fetch error / stale / offline phases would invent behaviour the
// web spec does not have (drift — Honesty Covenant §9). The branches the web source actually defines are this surface's
// complete state set, and each is projected here:
//   1. the per-option disabled state while a command is dispatching (web `disabled={loading}` + the `opacity-50
//      cursor-not-allowed` wash) — [isOptionEnabled],
//   2. the optional per-option description sub-line (web `opt.description && (...)`) — [visibleDescription] /
//      [hasDescription],
//   3. the option list itself (web `sc.options.map(...)`) — and its empty projection [isEmpty] / [hasOptions], which the
//      composable renders as a friendly empty state rather than the blank `space-y-2` box the web would collapse to when
//      a malformed `selectConfig` carries no options (a native polish that never hides the surface).
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/modals-dialogs/CommandSelectDialog — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package intentionally diverges
// from the path — exactly as the sibling modal/dialog surfaces do. `MatchingDeclarationName` is suppressed for the
// co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.modalsdialogs.commandselectdialog

import io.teslasync.shared.core.diagnostics.Logger

/**
 * One selectable command option — the native, already-localized mirror of the web `SelectOption`
 * (`features/system/commands.ts`) after `t(opt.labelKey, opt.labelFallback)` has resolved. The owning command surface
 * (which owns the `commands.ts` config + `useTranslation`) resolves [label] at the i18n boundary; [description] is the
 * raw, optional sub-line the web renders verbatim (web `opt.description`, never a key). [value] is the opaque token the
 * command dispatch uses and is what [CommandSelectDialog]'s `onSelect` hands back (web `onSelect(opt.value)`).
 *
 * @property value the parameter value submitted when this option is chosen (web `opt.value`).
 * @property label the localized option label shown as the option's primary line (web `t(opt.labelKey, …)`).
 * @property description an optional secondary descriptor; shown only when non-blank (web `opt.description`).
 */
data class CommandSelectOption(
    val value: String,
    val label: String,
    val description: String? = null,
)

/**
 * The pure derivations the composable renders over — the native mirror of the web component's inline `sc.options.map`,
 * the per-option `disabled={loading}` guard, and the `opt.description && …` conditional. Stateless and
 * side-effect-free, so it is fully covered by the off-device unit gate.
 */
object CommandSelectDialogProjection {
    /** Whether the picker has at least one option to show (web `sc.options.length > 0`). */
    fun hasOptions(options: List<CommandSelectOption>): Boolean = options.isNotEmpty()

    /**
     * Whether the picker resolved to no options — drives the friendly empty state instead of the blank `space-y-2`
     * box the web would render. A malformed/empty `selectConfig.options` is the only way this is true.
     */
    fun isEmpty(options: List<CommandSelectOption>): Boolean = options.isEmpty()

    /**
     * Whether an option button is actionable — the inverse of the web `disabled={loading}` (and the `opacity-50
     * cursor-not-allowed` wash). Every option disables together while a command dispatch is in flight.
     */
    fun isOptionEnabled(loading: Boolean): Boolean = !loading

    /**
     * Whether an option carries a visible description sub-line — the web `opt.description && (...)` truthiness. A
     * `null`, empty, or whitespace-only description is not shown.
     */
    fun hasDescription(option: CommandSelectOption): Boolean = !option.description.isNullOrBlank()

    /**
     * The description sub-line to render, or `null` when there is none — the native resolution of the web
     * `opt.description && (...)` conditional. Whitespace-only descriptions collapse to `null` so the sub-line is
     * never an empty row.
     */
    fun visibleDescription(option: CommandSelectOption): String? = option.description?.takeIf { it.isNotBlank() }
}

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object CommandSelectDialogRegistration {
    /** Stable surface id. */
    const val ID: String = "command-select-dialog"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "CommandSelectDialog"
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface
 * [CommandSelectDialogRegistration.SLUG] — never the command label, the option values/labels, or the chosen value — so
 * a diagnostics line can never leak which command the operator is configuring. Kept free of Compose so it is
 * unit-tested with a recording [Logger]; the composable calls it from its first-composition effect.
 */
object CommandSelectDialogDiagnostics {
    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to CommandSelectDialogRegistration.SLUG))
    }
}
